sql/core/src/main/scala/org/apache/spark/sql/execution/adaptive/AdaptiveSparkPlanExec.scala

```diff
@@ -536,7 +523,7 @@ case class AdaptiveSparkPlanExec(
       case s: ShuffleExchangeLike =>
         val newShuffle = applyPhysicalRules(
           s.withNewChildren(Seq(optimizedPlan)),
-          postStageCreationRules,
+          postStageCreationRules(outputColumnar = false),
```

Comment:
seems we don't need the post-stage creation extension rules at all if we have an API `ShuffleExchangeLike.requiresColumnarInput` (or reuse `SparkPlan.supportsColumnar`?). We can let the rule `ApplyColumnarRulesAndInsertTransitions` to adjust the plan and always output columnar format if the shuffle/broadcast needs columnar input.

Reply:
I am looking at this suggestion next to see how this would impact our usage. If we are to make changes like this I would prefer it if we could do these in a separate PR though.

Reply:
If that works without post-stage creation extension, I would prefer to revert it since Spark 3.2 is not out yet.

Reply:
I can try it with a new PR after this one is merged, and see if we can revert the post-stage creation extension.

Reply:
I think that we do need to retain the post-stage creation rules. Spark will insert a RowToColumnar in the adaptive plan for the final query stage but we need a way for our plugin to replace that with a GPU-specific version and we can use the extensible post-stage creation rule to do that.

Reply:
Can it be done with `ColumnarRule.postColumnarTransitions`?

Reply:
Yes, it can.